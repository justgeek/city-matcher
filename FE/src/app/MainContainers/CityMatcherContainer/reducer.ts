// actions.ts
import { isType, actionCreatorFactory } from "typescript-fsa";
import { Action } from "redux";
import {
  fetchingCities,
  fetchingOldCities,
  doActivateRow,
  doLogCity,
  doClearOldCities
} from "./actions";

export interface City {
  id: number;
  name: string;
  admin_area: string;
}
type State = {
  cities: City[];
  currentQueryOldCities: City[];
  currentActiveRow: number;
  currentLog: string;
};
export const initialState: State = {
  cities: [],
  currentQueryOldCities: [],
  currentActiveRow: null,
  currentLog: ""
};
const cityMatcherReducer = (state = initialState, action: Action): State => {
  if (isType(action, fetchingCities.done)) {
    const {
      payload: {
        result: { data }
      }
    } = action;
    return { ...state, cities: data };
  }

  if (isType(action, fetchingOldCities.done)) {
    const {
      payload: {
        result: { data }
      }
    } = action;
    return { ...state, currentQueryOldCities: data };
  }

  if (isType(action, doActivateRow)) {
    const {
      payload: { index }
    } = action;
    return { ...state, currentActiveRow: index };
  }

  if (isType(action, doLogCity)) {
    const {
      payload: { cityLog }
    } = action;
    return { ...state, currentLog: cityLog };
  }

  if (isType(action, doClearOldCities)) {
    return { ...state, currentQueryOldCities: [] };
  }

  return state;
};
export default cityMatcherReducer;
