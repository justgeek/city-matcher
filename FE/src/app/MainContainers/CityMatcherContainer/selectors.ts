import { State as AppState } from "../../Store/reducers";

export const selectCities = (state: AppState) => {
  return state.CityMatcher.cities;
};

export const selectCurrentAutoCompleteResults = (state: AppState) => {
  return state.CityMatcher.currentQueryOldCities;
};

export const selectCurrentActiveRow = (state: AppState) => {
  return state.CityMatcher.currentActiveRow;
};

export const selectCurrentLog = (state: AppState) => {
  return state.CityMatcher.currentLog;
};
