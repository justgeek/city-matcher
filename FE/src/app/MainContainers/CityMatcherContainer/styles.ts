import styled from "styled-components";

export const StyledCityMatcherContainer = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;
`;
