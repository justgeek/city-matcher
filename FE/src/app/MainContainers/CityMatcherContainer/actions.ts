import actionCreatorFactory from "typescript-fsa";
import { City } from "./reducer";
const actionCreator = actionCreatorFactory();

export interface fetchedCitiesPayload {
  data: City[];
}

export interface fetchOldCitiesPayload {
  query: string;
}

export interface activateRowPayload {
  index: number;
}

export interface logCityPayload {
  cityLog: string;
}
export const doFetchCities = actionCreator("FETCH_CITIES");
export const fetchingCities = actionCreator.async<{}, fetchedCitiesPayload>(
  "FETCHING_CITIES"
);
export const doFetchOldCities = actionCreator<fetchOldCitiesPayload>(
  "FETCH_OLD_CITIES"
);
export const fetchingOldCities = actionCreator.async<{}, fetchedCitiesPayload>(
  "FETCHING_OLD_CITIES"
);

export const doClearOldCities = actionCreator("CLEAR_CITIES");

export const doActivateRow = actionCreator<activateRowPayload>("ACTIVATE_ROW");

export const doLogCity = actionCreator<logCityPayload>("LOG_CITY");
