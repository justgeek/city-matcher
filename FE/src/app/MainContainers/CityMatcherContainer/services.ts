import { apiUrl } from "../../Config/api.config";
import { City } from "./reducer";

export class CityMatcherService {
  static async getCities(): Promise<City[]> {
    try {
      const response = await fetch(`${apiUrl}/v1/cities`);
      const result = await response.json();
      return result;
    } catch (e) {
      // Handle Error Here
    }
  }

  static async getOldCities(query): Promise<City[]> {
    try {
      const response = await fetch(`${apiUrl}/v1/autocomplete?q=${query}`);
      const result = await response.json();
      return result;
    } catch (e) {
      // Handle Error Here
    }
  }
}
