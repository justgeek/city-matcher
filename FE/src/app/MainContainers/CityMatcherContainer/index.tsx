import * as React from "react";
import { connect } from "react-redux";
import LoggerComponent from "./components/LoggerComponent";
import {
  doFetchCities,
  doFetchOldCities,
  fetchOldCitiesPayload,
  doActivateRow,
  doLogCity,
  doClearOldCities,
  logCityPayload
} from "./actions";
import { State as AppState } from "../../Store/reducers";
import {
  selectCities,
  selectCurrentAutoCompleteResults,
  selectCurrentActiveRow,
  selectCurrentLog
} from "./selectors";
import { debounce } from "throttle-debounce";
import CityMatcherListComponent from "./components/CityMatcherListComponent";
import { StyledCityMatcherContainer } from "./styles";

type Props = typeof mapDispatchToProps & ReturnType<typeof mapStateToProps>;

// Component DOM

class CityMatcherContainer extends React.PureComponent<Props> {
  constructor(props) {
    super(props);

    // Debounce search calls by 500 ms
    this.handleCityInputChange = debounce(500, this.handleCityInputChange);
  }
  componentDidMount() {
    const { doFetchCities } = this.props;

    doFetchCities();
  }

  private handleCityInputChange = ({ query }: fetchOldCitiesPayload) => {
    const { doFetchOldCities } = this.props;
    if (query) {
      doFetchOldCities({ query });
    }
  }

  private handleCityInputFocus = index => {
    const { doActivateRow, doClearOldCities } = this.props;
    // Clear current old cities so it doesn't appear on newly selected rows
    doClearOldCities();
    doActivateRow({ index });
  }

  private activateRow = (index: number) => {
    const { doActivateRow } = this.props;
    doActivateRow({ index });
  }

  private logCity = (cityLog: logCityPayload) => {
    const { doLogCity } = this.props;
    doLogCity(cityLog);
  }

  render() {
    const {
      cities,
      currentActiveRow,
      autoCompleteResults,
      currentLog
    } = this.props;
    return (
      <StyledCityMatcherContainer>
        <CityMatcherListComponent
          cities={cities}
          autoCompleteResults={autoCompleteResults}
          onCityInputChange={this.handleCityInputChange}
          onCityInputFocus={this.handleCityInputFocus}
          activateRow={this.activateRow}
          logCity={this.logCity}
          currentActiveRow={currentActiveRow}
        />
        <LoggerComponent currentLog={currentLog} />
      </StyledCityMatcherContainer>
    );
  }
}
const mapDispatchToProps = {
  doFetchCities,
  doFetchOldCities,
  doActivateRow,
  doLogCity,
  doClearOldCities
};
const mapStateToProps = (state: AppState) => {
  return {
    cities: selectCities(state),
    autoCompleteResults: selectCurrentAutoCompleteResults(state),
    currentActiveRow: selectCurrentActiveRow(state),
    currentLog: selectCurrentLog(state)
  };
};
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CityMatcherContainer);
