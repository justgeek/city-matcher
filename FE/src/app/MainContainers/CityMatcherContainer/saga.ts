import { takeLatest, call, put, takeEvery, take } from "redux-saga/effects";
import { Action } from "typescript-fsa";
import {
  doFetchCities,
  fetchingCities,
  doFetchOldCities,
  fetchingOldCities,
  fetchOldCitiesPayload
} from "./actions";
import { CityMatcherService } from "./services";

// watch for disptached action of type FETCH_CITIES
function* fetchCities() {
  yield put(fetchingCities.started({}));
  try {
    const result = yield call(CityMatcherService.getCities);
    yield put(fetchingCities.done({ params: {}, result: { data: result } }));
  } catch (e) {
    yield put(fetchingCities.failed({ params: {}, error: {} }));
  }
}

// watch for disptached action of type FETCH_OLD_CITIES
function* fetchOldCities(action: Action<fetchOldCitiesPayload>) {
  const {
    payload: { query }
  } = action;

  yield put(fetchingOldCities.started({}));
  try {
    const result = yield call(CityMatcherService.getOldCities, query);
    yield put(fetchingOldCities.done({ params: {}, result: { data: result } }));
  } catch (e) {
    yield put(fetchingOldCities.failed({ params: {}, error: {} }));
  }
}

export function* watchFetchCities() {
  yield takeEvery(doFetchCities, fetchCities);
}

export function* watchFetchOldCities() {
  yield takeEvery(doFetchOldCities, fetchOldCities);
}
export default [watchFetchCities, watchFetchOldCities];
