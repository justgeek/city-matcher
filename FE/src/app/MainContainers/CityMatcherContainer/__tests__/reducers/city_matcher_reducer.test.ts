import cityMatcherReducer, {
  initialState
} from "MainContainers/CityMatcherContainer/reducer";
import { doActivateRow } from "MainContainers/CityMatcherContainer/actions";

describe("INITIAL_STATE", () => {
  test("is initial", () => {
    const action = { type: "dummy_action" };

    expect(cityMatcherReducer(undefined, action)).toMatchSnapshot();
  });
});

describe("ACTIVE_ROW_STATE", () => {
  test("activate row by index returns expected state", () => {
    const index = 1;
    const action = { type: doActivateRow.type, payload: { index } };
    const expectedState = { ...initialState, currentActiveRow: index };
    expect(cityMatcherReducer(undefined, action)).toEqual(expectedState);
  });
});
