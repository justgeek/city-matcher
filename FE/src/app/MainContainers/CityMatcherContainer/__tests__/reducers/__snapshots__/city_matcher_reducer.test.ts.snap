// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`INITIAL_STATE is initial 1`] = `
Object {
  "cities": Array [],
  "currentActiveRow": null,
  "currentLog": "",
  "currentQueryOldCities": Array [],
}
`;
